### PR TITLE
[jest] Add type argument to jest.mock/doMock

### PR DESCRIPTION
**Pull request by pablobirukov**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

 _Rationale:_ WIth typescript 2.9+ it's possible to have module shape's type, as `typeof import("../foo")`. So type argument in `jest.mock` allows you to ensure that mock returned by `factory` has the same shape as mocked module:
```ts
    jest.mock<typeof import("../foo")>("../foo", () => ({ /*  */}));
```

### CODE REVIEW

**Review by r3nya — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Not sure how I feel about this; why can't you add a type annotation to the function itself?

**Comment by r00ger** [comment]
@DanielRosenwasser It definitely can be enforced by defining a return type of a function.

To be frank, it was simpler to create a fixer for tslint rule that enforces type argument rather the one that ensures return type (e.g. when we pass a reference to `factory` instead of inlining it: `jest.mock("", factory)`), that's why I decided to go this way, and the purpose of the PR was to hear the opinion of the community.

On one hand, it still seems right to me, since it enforces typesafety (in a way), on the other hand, the need for type argument might not be obvious for library consumers

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -126,7 +126,7 @@ declare namespace jest {
     /**
      * Mocks a module with an auto-mocked version when it is being required.
      */
-    function mock(moduleName: string, factory?: any, options?: MockOptions): typeof jest;
+    function mock<T extends {}>(moduleName: string, factory?: () => T, options?: MockOptions): typeof jest;
```
Comment (Kovensky) [inline types/jest/index.d.ts]: Would

```ts
function mock<T extends string>(moduleName: T, factory?: () => typeof import(T), options?: MockOptions): this
```

work?

Comment (andys8) [reply types/jest/index.d.ts]: `factory?: () => typeof import(T)` doesn't look to make sense. Have you seen it being used that way before?

Comment (r00ger) [reply types/jest/index.d.ts]: It won't work, the syntax is incorrect. `import` expect string literal as argument

**Comment by minestarks** [comment]
I'm also hesitant to merge this. It doesn't actually enforce type safety unless the user thinks to add the type argument -- and if the user thinks of it there is already a more obvious way of doing it (adding the type to the function itself).

**Review by minestarks — CHANGES_REQUESTED** [review]
Putting PR into the "request changes" state since I'm not convinced we should take this PR at this point

**Pull request closed without merging**